cni-plugin/deployment/scripts/install-cni.sh

```diff
@@ -163,58 +202,148 @@ contexts:
+# plugin's configuration file.
+sync() {
+  local filename="$1"
+  local ev="$2"
+  local filepath="${HOST_CNI_NET}/$filename"
+
+  local prev_sha="$3"
+
+  local config_file_count
+  local new_sha
+  if [ "$ev" = "DELETE" ]; then
+    # When the event type is 'DELETE', we check to see if there are any `*conf` or `*conflist`
+    # files on the host's filesystem. If none are present, we install in
+    # 'interface' mode, using our own CNI config file.
+    config_file_count=$(find "${HOST_CNI_NET}" -maxdepth 1 -type f \( -iname '*conflist' -o -iname '*conf' \) | sort | wc -l)
+    if [ "$config_file_count" -eq 0 ]; then
+      echo "No active CNI configuration file found after $ev event; re-installing in \"interface\" mode"
+      install_cni_conf "${DEFAULT_CNI_CONF_PATH}"
+    fi
+  elif [ "$ev" = "CREATE" ]; then
+    # When the event type is 'CREATE', we check the previously observed SHA (updated
+    # with each file watch) and compare it against the new file's SHA. If they
+    # differ, it means something has changed.
+    new_sha=$(sha256sum "${filepath}" | awk '{print $1}')
+    if [ "$new_sha" != "$prev_sha" ]; then
+      # Create but don't rm old one since we don't know if this will be configured
+      # to run as _the_ cni plugin.
+      echo "New file [$filename] detected; re-installing in \"chained\" mode"
+      install_cni_conf "$filepath"
+    fi
+  else
+    # If the SHA hasn't changed or we get an unrecognised event, ignore it.
+    # When the SHA is the same, we can get into infinite loops whereby a file has
+    # been created and after re-install the watch keeps triggering CREATE events
+    # that never end.
+    echo "Ignoring event: $ev $filepath; no real changes detected"
+  fi
+}
+
+# Monitor will start a watch on host's CNI config directory. Although files are
+# mostly `mv'd`, because they are moved from the container's filesystem, the
+# events logged will typically be a DELETED followed by a CREATE. When we are on
+# the same system partition, `mv` simply renames, however, that won't be the
+# case so we don't watch any "moved_to" or "moved_from" events.
+monitor() {
+  inotifywait -m "${HOST_CNI_NET}" -e create,delete |
+    while read -r directory action filename; do
+      if [[ "$filename" =~ .*.(conflist|conf)$ ]]; then 
+        echo "Detected change in $directory: $action $filename"
+        sync "$filename" "$action" "$cni_conf_sha"
+        # When file exists (i.e we didn't deal with a DELETE ev)
+        # then calculate its sha to be used the next turn.
+        if [ -e "$directory/$filename" ]; then
+          cni_conf_sha="$(sha256sum "$directory/$filename" | awk '{print $1}')"
+        fi
+      fi
+    done
+}
+
+################################
+### CNI Plugin Install Logic ###
+################################
+
+install_cni_bin
+# Install for the first time using whatever path we have chosen at the start.
+install_cni_conf "${CNI_CONF_PATH}"
```

Comment:
if I'm reading this correctly, on startup we install into the lexicographically first conf file only.  This means that if there are other conf files but they never change, inotifywait will never get triggered for them and we will never install into them.  Is this the desired behavior?  Should we install into all conf files right off the bat?

Reply:
Yeah, you're right, we should make sure we're consistent with the rest of our logic. I'll change this.

Reply:
Done and tested

<details><summary> spoiler: logs </summary>

```text
# Install our CNI
Wrote linkerd CNI binaries to /host/opt/cni/bin
No active CNI configuration files found; installing in "interface" mode in /host/etc/cni/net.d/01-linkerd-cni.conf

# Install calico, expect changes from watch
Removing Linkerd's configuration file: /host/etc/cni/net.d/01-linkerd-cni.conf
Created CNI config /host/etc/cni/net.d/10-calico.conflist
Detected change in /host/etc/cni/net.d/: DELETE 01-linkerd-cni.conf
Detected change in /host/etc/cni/net.d/: DELETE 10-calico.conflist
Detected change in /host/etc/cni/net.d/: CREATE 10-calico.conflist
Ignoring event: CREATE /host/etc/cni/net.d/10-calico.conflist; no real changes detected

# Delete linkerd-cni, re-install, expect to append
```
Wrote linkerd CNI binaries to /host/opt/cni/bin
Installing CNI configuration in "chained" mode for /host/etc/cni/net.d/10-calico.conflist
```
</details>